blockindex/indexer.go

```diff
@@ -0,0 +1,425 @@
+// Copyright (c) 2019 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package blockindex
+
+import (
+	"bytes"
+	"context"
+	"math/big"
+	"sync"
+
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/db"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+)
+
+// the NS/bucket name here are used in index.db, which is separate from chain.db
+// still we use 2-byte NS/bucket name here, to clearly differentiate from those (3-byte) in BlockDAO
+const (
+	// first 12-byte of hash is cut off, only last 20-byte is written to DB to reduce storage
+	hashOffset          = 12
+	blockHashToHeightNS = "hh"
+	actionToBlockHashNS = "ab"
+)
+
+var (
+	totalBlocksBucket  = []byte("bk")
+	totalActionsBucket = []byte("ac")
+	// ErrActionIndexNA indicates action index is not supported
+	ErrActionIndexNA = errors.New("action index not supported")
+)
+
+type (
+	addrIndex map[hash.Hash160]db.CountingIndex
+
+	// Indexer is the interface for block indexer
+	Indexer interface {
```

Comment:
we can move the definition into blockchain/
Indexer interface could be as simple as
```
    Indexer interface {
      Start(context.Context) error
      Stop(context.Context) error
      PutBlock(*block.Block) error
      DeleteBlock(*block.Block) error
      Commit() error
    }
```